### PR TITLE
Add BIPOP-CMA-ES support in `CmaEsSampler`

### PR DESCRIPTION
**Pull request by cross32768**
## Motivation
Current `CmaEsSampler` has IPOP-CMA-ES as the restart strategy.
[BIPOP-CMA-ES](https://hal.inria.fr/hal-00818596v1/document) is a kind of extension of IPOP-CMA-ES to improve performance in some problems by adding device about the increment of population size.
I want to consider adding BIPOP-CMA-ES as new option.

## Description of the changes

* Add BIPOP-CMA-ES option in `optuna/samplers/_cmaes.py` . Users can use this new feature by `optuna.samplers.CmaEsSampler(restart_strategy="bipop")`
  * Add population increment logic of BIPOP-CMA-ES in `sampler_relative` method of `CmaEsSampler`
  * Add some system_attrs to manage counter variable required for BIPOP-CMA-ES
  * Change the return format of `_restore_optimizer` method of `CmaEsSampler` and add code to load counter variable like `n_restarts` other place
  * Change the instance variable name `_popsize` to `_initial_popsize` because its more adequate
  * Add comments and docstring to describe about BIPOP-CMA-ES
* Add test for BIPOP-CMA-ES in `tests/sampler_tests/test_cmaes.py`

### CODE REVIEW

**Comment by cross32768** [comment]
@c-bata Could you review this PR?

**Comment by cross32768** [comment]
This gif shows behavior of BIPOP-CMA-ES on himmelblau function.
bipop_himmelblau

**Comment by HideakiImamura** [comment]
@Alnusjaponica Could you review this PR?

**Review by Alnusjaponica — COMMENTED** [review]
Sorry for delayed reply. I made some small comments mainly for code comments. If you don't agree with them, please disregard them.

File: optuna/samplers/_cmaes.py [hunk 1]
```diff
@@ -408,6 +439,30 @@ def sample_relative(
...
+                if poptype == "small":
+                    small_n_eval += n_eval
+                else:  # poptype == "large"
+                    large_n_eval += n_eval
+
+                if small_n_eval < large_n_eval:
+                    poptype = "small"
+                    popsize_multiplier = self._inc_popsize**n_restarts_with_large
+                    popsize = math.floor(
+                        self._initial_popsize * popsize_multiplier ** (np.random.uniform() ** 2)
```
[the remark this answers is not in the document]
Comment (cross32768) [reply optuna/samplers/_cmaes.py:455]: Thank you. It's my mistake. I'll fix it.

File: optuna/samplers/_cmaes.py [hunk 2]
```diff
@@ -153,6 +161,8 @@ def objective(trial):
             Strategy for restarting CMA-ES optimization when converges to a local minimum.
             If given :obj:`None`, CMA-ES will not restart (default).
             If given 'ipop', CMA-ES will restart with increasing population size.
+            if given 'bipop', CMA-ES will restart with occasionally increasing
+            population size or small population size.
```
[the remark this answers is not in the document]
Comment (cross32768) [reply optuna/samplers/_cmaes.py:165]: Yes. My comment is not appropriate. I'll fix it

**Comment by cross32768** [comment]
@Alnusjaponica Thanks you for comments. I fixed all points.

File: optuna/samplers/_cmaes.py [hunk 3]
```diff
@@ -92,6 +97,9 @@ def objective(trial):
       size. In Proceedings of the IEEE Congress on Evolutionary Computation (CEC 2005),
       pages 1769–1776. IEEE Press, 2005.
       <http://www.cmap.polytechnique.fr/~nikolaus.hansen/cec2005ipopcmaes.pdf>`_
+    - `N. Hansen. Benchmarking a BI-Population CMA-ES on the BBOB-2009 Function Testbed.
+      GECCO Workshop, 2009.
+      <https://hal.inria.fr/inria-00382093/document>`_
```
[the remark this answers is not in the document]
Comment (cross32768) [reply optuna/samplers/_cmaes.py]: Thank you for pointing this out. I'll fix it. As a related topic, BIPOP-CMA-ES has 2 papers ("BI-population CMA-ES Algorithms with Surrogate Models and Line Searches" and "Benchmarking a BI-Population CMA-ES on the BBOB-2009 Function Testbed" ). I'm not sure which one to quote, what do you think @c-bata ?

[the remark this answers is not in the document]
Comment (c-bata) [reply optuna/samplers/_cmaes.py]: @nomuramasahir0 Can I ask you about this? It seems that https://github.com/CyberAgentAILab/cmaes also quotes the previous one.

[the remark this answers is not in the document]
Comment (nomuramasahir0) [reply optuna/samplers/_cmaes.py]: I think that quoting "Benchmarking a BI-Population CMA-ES on the BBOB-2009 Function Testbed" is better, as the paper "BI-population CMA-ES Algorithms with Surrogate Models and Line Searches" investigates BIPOP-CMA with additional operations, which are not included in the https://github.com/CyberAgentAILab/cmaes.

File: optuna/samplers/_cmaes.py [hunk 4]
```diff
@@ -429,9 +484,22 @@ def sample_relative(
         study._storage.set_trial_system_attr(
             trial._trial_id, generation_attr_key, optimizer.generation
         )
+        study._storage.set_trial_system_attr(
+            trial._trial_id, self._attr_keys.popsize, optimizer.population_size
+        )
```
Comment (c-bata) [inline optuna/samplers/_cmaes.py:489]: I am wondering if the following code will work correctly without saving a `restart_strategy` in `system_attrs`.

```python
import math
import optuna


def objective(trial: optuna.Trial) -> float:
    return sum([
        math.sin(trial.suggest_float('param-{}'.format(i), -10 * i, 10 * i))
        for i in range(10)
    ])


if __name__ == '__main__':
    storage = optuna.storages.InMemoryStorage()

    ipop = optuna.samplers.CmaEsSampler(restart_strategy="ipop")
    study = optuna.create_study(sampler=ipop, storage=storage, study_name="bipop-example")
    study.optimize(objective, n_trials=100)

    bipop = optuna.samplers.CmaEsSampler(restart_strategy="bipop")
    study = optuna.load_study(sampler=bipop, storage=storage, study_name="bipop-example")
    study.optimize(objective, n_trials=100)

    print(study.best_trial)
```

More precisely, when the restart strategy changes from ipop to bipop, do you think the parameters stored in system_attr such as n_restarts and popsize should be reused?

**Review by c-bata — COMMENTED** [review]
Thank you for your pull request! At a glance, changes basically look good to me.
Left an early comment since I noticed while quickly taking a look. I will thoroughly review this PR tomorrow.

**Review by c-bata — COMMENTED** [review]
I confirmed that BIPOP-CMA-ES works as expected.

Screenshot 2023-03-16 15 26 12

```python
import optuna
from optuna_dashboard import run_server


def objective(trial: optuna.Trial) -> float:
    x = trial.suggest_float('x', -10, 10)
    y = trial.suggest_float('y', -10, 10)
    system_attrs = trial.storage.get_trial_system_attrs(trial._trial_id)
    for key, value in system_attrs.items():
        if "optimizer" in key:
            trial.set_user_attr(key, f"{value[:5]}...")
        else:
            trial.set_user_attr(key, value)
    return x ** 2 + y **2


if __name__ == '__main__':
    storage = optuna.storages.InMemoryStorage()

    sampler = optuna.samplers.CmaEsSampler(restart_strategy="bipop")
    study = optuna.create_study(sampler=sampler, storage=storage, study_name="bipop-example")
    study.optimize(objective, n_trials=200)

    print(study.best_trial)
    run_server(storage, port=8080)
```

However, as I commented yesterday, system_attrs must be separated by `restart_strategy`. Otherwise, it won't work correctly after switching the `restart_strategy` from ipop to bipop.

**Comment by Alnusjaponica** [comment]
Sorry for my delayed reply. Thank you for your update! LGTM.

**Review by Alnusjaponica — APPROVED** [review]

**Comment by Alnusjaponica** [comment]
This is trifle but there's a typo in tests/samplers_tests/test_cmaes.py:327 and 
```
# Test sampler attr_key propery
```
should be 
```
# Test sampler attr_key property.
```

File: optuna/samplers/_cmaes.py (hunk 4 above)
Comment (cross32768) [reply optuna/samplers/_cmaes.py]: I update code to deal with this problem. Could you take another look?

File: tests/samplers_tests/test_cmaes.py [hunk 5]
```diff
@@ -324,19 +327,25 @@ def _create_trials() -> List[FrozenTrial]:
     ],
 )
 def test_sampler_attr_key(options: Dict[str, bool], key: str) -> None:
-    # Test sampler attr_key propery
+    # Test sampler attr_key propery.
```
Comment (Alnusjaponica) [inline tests/samplers_tests/test_cmaes.py:330]:
```suggestion
    # Test sampler attr_key property.
```
I guess what is meant here is 'property'

Comment (cross32768) [reply tests/samplers_tests/test_cmaes.py]: Sorry I forgot to fix this typo. I fixed now

File: optuna/samplers/_cmaes.py [hunk 6]
```diff
@@ -369,10 +380,46 @@ def sample_relative(
...
+        store_optimizer = False
+        optimizer = self._restore_optimizer(completed_trials)
+        optimizer_inited = False
         if optimizer is None:
-            n_restarts = 0
-            optimizer = self._init_optimizer(trans, study.direction, population_size=self._popsize)
+            optimizer = self._init_optimizer(
+                trans, study.direction, population_size=self._initial_popsize
+            )
+            optimizer_inited = True
```
Comment (c-bata) [inline optuna/samplers/_cmaes.py:393]: I am currently looking for a way to make this logic simpler. If we need to reinitialize the optimizer for different options, how about having a different system_attr to store the optimizer for each option?

Specifically, there is a way to store the optimizer in the key obtained by the following function.

```diff
diff --git a/optuna/samplers/_cmaes.py b/optuna/samplers/_cmaes.py
index a8731773f..6f1949465 100644
--- a/optuna/samplers/_cmaes.py
+++ b/optuna/samplers/_cmaes.py
@@ -46,7 +46,7 @@ _SYSTEM_ATTR_MAX_LENGTH = 2045


 class _CmaEsAttrKeys(NamedTuple):
-    optimizer: str
+    optimizer: Callable[[int], str]
     generation: Callable[[int], str]
     popsize: Callable[[], str]
     n_restarts: Callable[[], str]
@@ -534,6 +534,9 @@ class CmaEsSampler(BaseSampler):
         else:
             attr_prefix = "cma:"

+        def optimizer_key_template(restart: int) -> str:
+            return attr_prefix + f"{self._restart_strategy}:n_restart_{restart}:optimizer"
+
         def generation_attr_key_template(restart: int) -> str:
             if self._restart_strategy is None:
                 return attr_prefix + "generation"
@@ -555,7 +558,7 @@ class CmaEsSampler(BaseSampler):
                 return attr_prefix + "{}:n_restarts".format(self._restart_strategy)

         return _CmaEsAttrKeys(
-            attr_prefix + "optimizer",
+            optimizer_key_template,
             generation_attr_key_template,
             popsize_attr_key_template,
             n_restarts_attr_key_template,
```

What do you think?

**Review by c-bata — COMMENTED** [review]
Sorry for the delay in my review. It works fine, but there seems to be room for improvement on how to restore the optimizer. I left one suggestion.

File: optuna/samplers/_cmaes.py (hunk 6 above)
Comment (cross32768) [reply optuna/samplers/_cmaes.py:393]: Ok, I will try this approach.

Comment (cross32768) [reply optuna/samplers/_cmaes.py]: I updated code by applying this approach and it looks works well. Could you take another look?

Comment (cross32768) [reply optuna/samplers/_cmaes.py]: @c-bata I updated code by applying this approach and it looks works well. Could you take another look?

**Review by c-bata — APPROVED** [review]
Thanks for the update! The update looks great to me.

Just for the record, I confirmed that BIPOP-CMA-ES does not take over the optimization history of IPOP-CMA-ES with the following Python script.

```python
import optuna
from optuna_dashboard import run_server


seed = 1
storage = optuna.storages.InMemoryStorage()


def objective(trial: optuna.Trial) -> float:
    x = trial.suggest_float('x', -10, 10)
    y = trial.suggest_float('y', -10, 10)
    system_attrs = trial.storage.get_trial_system_attrs(trial._trial_id)
    for key, value in system_attrs.items():
        if "optimizer" in key:
            trial.set_user_attr(key, f"{value[:5]}...")
        else:
            trial.set_user_attr(key, value)
    return x ** 2 + y ** 2


if __name__ == '__main__':

    study1 = optuna.create_study(sampler=optuna.samplers.RandomSampler(seed=seed), storage=storage, study_name="random-to-bipop")
    study1.optimize(objective, n_trials=100)
    study1 = optuna.load_study(sampler=optuna.samplers.CmaEsSampler(restart_strategy="bipop", seed=seed), storage=storage, study_name="random-to-bipop")
    study1.optimize(objective, n_trials=100)

    study2 = optuna.create_study(sampler=optuna.samplers.CmaEsSampler(restart_strategy="ipop", seed=seed), storage=storage, study_name="ipop-to-bipop")
    study2.optimize(objective, n_trials=100)
    study2 = optuna.load_study(sampler=optuna.samplers.CmaEsSampler(restart_strategy="bipop", seed=seed), storage=storage, study_name="ipop-to-bipop")
    study2.optimize(objective, n_trials=100)

    for t1, t2 in zip(study1.trials[100:], study2.trials[100:]):
        assert t1.params == t2.params and t1.value == t2.value
    run_server(storage, port=8080)
```

**Pull request merged**